File: android/app/src/main/java/ai/thetahealth/mirobody/data/llm/OnDeviceLlmEngine.kt
package ai.thetahealth.mirobody.data.llm

import ai.thetahealth.mirobody.data.chat.dto.ChatStreamEvent
import kotlinx.coroutines.flow.Flow

/** One settled turn of a conversation, fed to the on-device model as context. */
data class ChatTurn(val fromUser: Boolean, val text: String)

/**
 * A locally-hosted LLM that produces the same [ChatStreamEvent] stream the remote
 * `/api/chat` SSE path does, so the existing chat UI renders on-device replies with
 * no UI changes. Implementations run fully on-device — no network, no server.
 *
 * The Android implementation is [LiteRtLlmEngine] (Gemma 4 via LiteRT-LM). The
 * interface is deliberately runtime-agnostic so an ML Kit GenAI (Gemini Nano) path
 * can slot in later behind the same seam.
 */
interface OnDeviceLlmEngine {
    /**
     * Stream a reply for [history]; the final entry is the new user turn. Emits
     * [ChatStreamEvent.Reply] deltas (and optional [ChatStreamEvent.Thinking]),
     * then [ChatStreamEvent.End]. Failures surface as [ChatStreamEvent.Error]
     * rather than throwing, mirroring the SSE client's contract.
     */
    fun generate(history: List<ChatTurn>): Flow<ChatStreamEvent>

    /** Release native resources (the model file stays on disk). Safe to call repeatedly. */
    fun close()
}

File: ios/Mirobody/Data/LLM/OnDeviceLlmEngine.swift
import Foundation

/// One settled turn of a conversation, fed to the on-device model as context.
/// Mirrors Android's `data/llm/OnDeviceLlmEngine.kt` `ChatTurn`.
struct ChatTurn {
    let fromUser: Bool
    let text: String
}

/// A locally-hosted LLM that produces the same `ChatStreamEvent` stream the remote
/// `/api/chat` SSE path does, so the existing chat UI renders on-device replies with
/// no UI changes. Runs fully on-device — no network, no server.
///
/// The concrete implementation is `LiteRtLlmEngine` (Gemma 4 via LiteRT-LM). The
/// protocol is runtime-agnostic, matching the Android `OnDeviceLlmEngine` seam.
protocol OnDeviceLlmEngine: AnyObject {
    /// Stream a reply for `history`; the final entry is the new user turn. Emits
    /// `.reply` deltas (and optional `.thinking`), then `.end`. Failures surface as
    /// `.error` rather than throwing, mirroring the SSE client's contract.
    func generate(history: [ChatTurn]) -> AsyncStream<ChatStreamEvent>

    /// Release native resources (the model file stays on disk). Safe to call repeatedly.
    func close()
}
